ts4.1/index.d.ts

```diff
@@ -0,0 +1,294 @@
+	 * // => "fooBar" | "bar_baz"
+	 *
+	 * type T1 = CamelcaseUnless<"foo_bar" | "bar_baz", false, string>;
+	 * // => "foo_bar" | "fooBar" | "bar_baz" | "barBaz"
+	 * ```
+	 */
+	type CamelcaseUnless<S extends string, PascalCase extends boolean, Exclude extends string> =
+		// We treat `string` as "indeterminate set of strings", so return both cases in this case.
+		string extends Exclude ? S | Camelize<S, PascalCase> :
+		S extends Exclude ? S : Camelize<S, PascalCase>;
+
+	/**
+	 * @internal
+	 * Compute return value of camelcaseKeys.
+	 *
+	 * @param T an object or array to camelcase its keys
+	 * @param PascalCase capitalize the first word
+	 * @param Deep recursively camelcase
+	 * @param Exclude don't camelcase these keys
+	 * @param StopPaths don't recursive into these paths
+	 */
+	type CamelcaseKeys<
+		T,
+		PascalCase extends boolean = false,
+		Deep extends boolean = false,
+		Exclude extends string = never,
+		StopPaths extends string = never,
+	> =
+		// Array case
+		T extends ReadonlyArray<any> ?
+		{ [K in keyof T]: CamelcaseKeys<T[K], PascalCase, Deep, Exclude, StopPaths> } :
+		// Object case
+		T extends Record<string, any> ?
+		{
+			[K in keyof T as CamelcaseUnless<K & string, PascalCase, Exclude>]:
+				Deep extends true ?
+				K extends StopPaths ?
+				T[K] :
+				CamelcaseKeys<T[K], PascalCase, Deep, Exclude, FilterPaths<StopPaths, K & string>> :
+				T[K];
+		} :
+		// Otherwise (e.g. number, string, null, undefined)
+		T;
+
+	/**
+	 * @internal
+	 * Helper type for `ExtractExcludes`. Used to turn `RegExp` into `string`.
+	 */
+	type ExtractExclude<T> = T extends string ? T : string;
+	/**
+	 * @internal
+	 * Turn the `exclude` argument into a union type, which can be pased to `DeepCamelcaseKeys`.
+	 *
+	 * @param T an array of string or RegExp
+	 *
+	 * @example
+	 * ```
+	 * type T1 = ExtractExcludes<["foo", "bar"]>;
+	 * // => "foo" | "bar"
+	 * type T2 = ExtractExclude<["foo", "bar", string]>;
+	 * // => string
+	 * type T3 = ExtractExclude<["foo", "bar", RegExp]>;
+	 * // => string
+	 * ```
+	 */
+	type ExtractExcludes<T extends ReadonlyArray<string | RegExp>> = { [K in keyof T]: ExtractExclude<T[K]> }[number];
+	/**
+	 * @internal
+	 * Turn null/undefined to another type. Helper for `DefaultValue`.
+	 */
+	type Default<T, U> = T extends null | undefined ? U : T;
+	/**
+	 * @internal
+	 * Pick property type if it exists, fallback to default otherwise.
+	 *
+	 * @param T a record type
+	 * @param K a key
+	 * @param U a fallback type
+	 *
+	 * @example
+	 * ```
+	 * type T1 = DefaultValue<{ a: number }, "a", string>;
+	 * // => number
+	 * type T2 = DefaultValue<{ a?: number }, "a", string>;
+	 * // => number | string
+	 * type T3 = DefaultValue<{ a: undefined }, "a", string>;
+	 * // => string
+	 * type T4 = DefaultValue<{}, "a", string>;
+	 * // => string
+	 * ```
+	 */
+	type DefaultValue<T, K extends string, U> = T extends { [key in K]?: any } ? Default<T[K], U> : U;
```

Comment:
```suggestion
	type PickWithDefault<T, K extends string, U> = T extends { [key in K]?: any } ? Default<T[K], U> : U;
```

or `PickWithFallback`?